src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -494,18 +585,56 @@ int encryptFinal(byte[] in, int inOfs, int len, byte[] out, int outOfs)
             doLastBlock(in, inOfs, len, out, outOfs, true);
         }
 
-        byte[] lengthBlock =
-            getLengthBlock(sizeOfAAD, processed);
-        ghashAllToS.update(lengthBlock);
+        byte[] block = getLengthBlock(sizeOfAAD, processed);
+        ghashAllToS.update(block);
         byte[] s = ghashAllToS.digest();
-        byte[] sOut = new byte[s.length];
+        if (tagLenBytes > block.length) {
+            block = new byte[tagLenBytes];
+        }
         GCTR gctrForSToTag = new GCTR(embeddedCipher, this.preCounterBlock);
-        gctrForSToTag.doFinal(s, 0, s.length, sOut, 0);
+        gctrForSToTag.doFinal(s, 0, s.length, block, 0);
 
-        System.arraycopy(sOut, 0, out, (outOfs + len), tagLenBytes);
+        System.arraycopy(block, 0, out, (outOfs + len), tagLenBytes);
         return (len + tagLenBytes);
     }
 
+    int encryptFinal(ByteBuffer src, ByteBuffer dst)
+        throws IllegalBlockSizeException, ShortBufferException {
+        int len = src.remaining();
+        dst.mark();
+        if (len > MAX_BUF_SIZE - tagLenBytes) {
+            throw new ShortBufferException
+                ("Can't fit both data and tag into one buffer");
+        }
+
+        if (dst.remaining() < len + tagLenBytes) {
+            throw new ShortBufferException("Output buffer too small");
+        }
+
+        checkDataLength(processed, len);
+
+        processAAD();
+        if (len > 0) {
+            doLastBlock((ibuffer == null || ibuffer.size() == 0) ?
+                    null : ByteBuffer.wrap(ibuffer.toByteArray()), src, dst);
+            dst.reset();
+            ghashAllToS.doLastBlock(dst, processed);
+        }
+
+        byte[] block = getLengthBlock(sizeOfAAD, processed);
+        ghashAllToS.update(block);
+
+        byte[] s = ghashAllToS.digest();
+        if (tagLenBytes > block.length) {
+            block = new byte[tagLenBytes];
+        }
+        GCTR gctrForSToTag = new GCTR(embeddedCipher, this.preCounterBlock);
+        gctrForSToTag.doFinal(s, 0, s.length, block, 0);
+        dst.put(block, 0, tagLenBytes);
+
+        return (processed + tagLenBytes);
```

Comment:
Is it supposed to return "all data processed + tag len"? Normally, we return the output size produced by this particular call instead of all accumulated.

Reply:
Yes, that should be len + tagLenBytes.  We obviously don't check that in our tests